File: main-agent/src/main/java/com/firefly/main/inbox/Inbox.java
package com.firefly.main.inbox;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.FileWriter;
import java.io.RandomAccessFile;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * 持久收件箱（teammate 通信，.jsonl）
 * /data/inbox/{agent_name}.jsonl，append 写 / drain 读
 * 用于 deploy_subagent 回报灰度状态、rollback_subagent 回报退化等跨轮通信
 * 源自乐享 s09/s15 MessageBus 机制
 */
@Component
public class Inbox {

    private final ObjectMapper mapper = new ObjectMapper();
    private static final String INBOX_DIR = System.getenv().getOrDefault("INBOX_DIR", "/data/inbox");

    /** 写消息（append 模式，跨重启不丢） */
    public void send(String toAgent, Map<String, Object> message) {
        try {
            File dir = new File(INBOX_DIR);
            if (!dir.exists()) dir.mkdirs();
            File f = new File(dir, toAgent + ".jsonl");
            message.put("ts", System.currentTimeMillis() / 1000);
            try (FileWriter fw = new FileWriter(f, true)) {
                fw.write(mapper.writeValueAsString(message) + "\n");
            }
        } catch (Exception e) {
            throw new RuntimeException("收件箱写入失败", e);
        }
    }

    /** 读取并清空所有消息（drain） */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> drain(String agentName) {
        List<Map<String, Object>> msgs = new ArrayList<>();
        File f = new File(INBOX_DIR, agentName + ".jsonl");
        if (!f.exists()) return msgs;
        try (RandomAccessFile raf = new RandomAccessFile(f, "rw")) {
            String line;
            while ((line = raf.readLine()) != null) {
                if (line.trim().isEmpty()) continue;
                msgs.add(mapper.readValue(line.getBytes(StandardCharsets.UTF_8), Map.class));
            }
            // 清空文件
            raf.setLength(0);
        } catch (Exception e) {
            throw new RuntimeException("收件箱读取失败", e);
        }
        return msgs;
    }
}

File: legacy/prototype-v0/scout-agent/src/inbox/inbox.py
"""
持久收件箱（三 Agent 共用，teammate 通信，.jsonl）
/data/inbox/{agent_name}.jsonl，append 写 / drain 读
跨重启不丢，用于 deploy_subagent/rollback_subagent 回报等跨轮通信
源自乐享 s09/s15 MessageBus
"""
import os
import json
from typing import List, Dict


class Inbox:
    INBOX_DIR = os.getenv("INBOX_DIR", "/data/inbox")

    def send(self, to_agent: str, message: Dict):
        """写消息（append，跨重启不丢）"""
        os.makedirs(self.INBOX_DIR, exist_ok=True)
        path = os.path.join(self.INBOX_DIR, f"{to_agent}.jsonl")
        message["ts"] = int(__import__("time").time())
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(message, ensure_ascii=False) + "\n")

    def drain(self, agent_name: str) -> List[Dict]:
        """读取并清空所有消息"""
        path = os.path.join(self.INBOX_DIR, f"{agent_name}.jsonl")
        if not os.path.exists(path):
            return []
        with open(path, "r+", encoding="utf-8") as f:
            lines = f.readlines()
            f.seek(0); f.truncate()  # 清空
        msgs = []
        for line in lines:
            line = line.strip()
            if line:
                msgs.append(json.loads(line))
        return msgs
